File: erp-backend/src/main/java/com/erp/controller/ShopController.java
package com.erp.controller;

import com.erp.entity.Shop;
import com.erp.repository.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Sort;
import org.springframework.http.ResponseEntity;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
public class ShopController {

    @Autowired
    private ShopRepository shopRepository;
    
    @Autowired
    private OrderRepository orderRepository;

    @Autowired
    private OrderItemRepository orderItemRepository;

    @Autowired
    private LedgerRepository ledgerRepository;

    @Autowired
    private UserRepository userRepository;

    // POST: Nayi dukan register karna — Admin only
    @PostMapping("/add-shop")
    public ResponseEntity<?> addShop(
            @RequestAttribute("role") String role,
            @RequestBody Map<String, String> body) {
        // ✅ SECURITY FIX: Only admin can add shops
        if (!"admin".equalsIgnoreCase(role)) {
            return ResponseEntity.status(403).body("Access Denied: Only Admin can add shops");
        }
        Shop shop = new Shop();
        shop.setShopName(body.get("shop_name"));
        shop.setShopAddress(body.get("shop_address"));
        shop.setTotalDebt(java.math.BigDecimal.ZERO);
        return ResponseEntity.ok(shopRepository.save(shop));
    }

    // GET: Saare shops ki list — Admin and Staff
    @GetMapping("/shops")
    public ResponseEntity<?> getAllShops(@RequestAttribute("role") String role) {
        // ✅ SECURITY FIX: Only admin or staff can list all shops
        if (!"admin".equalsIgnoreCase(role) && !"staff".equalsIgnoreCase(role) && !"shopkeeper".equalsIgnoreCase(role)) {
            return ResponseEntity.status(403).body("Access Denied: Unauthorized");
        }
        return ResponseEntity.ok(shopRepository.findAll(Sort.by(Sort.Direction.ASC, "shopId")));
    }

    // DELETE: Shop delete karna — Admin only
    @Transactional
    @DeleteMapping("/delete-shop/{id}")
    public ResponseEntity<String> deleteShop(
            @RequestAttribute("role") String role,
            @PathVariable Integer id) {
        // ✅ SECURITY FIX: Only admin can delete shops
        if (!"admin".equalsIgnoreCase(role)) {
            return ResponseEntity.status(403).body("Access Denied: Only Admin can delete shops");
        }
        try {
            // 1. Order Items delete karo (JO orders is shop se linked hain)
            orderItemRepository.deleteByShopId(id);
            
            // 2. Orders delete karo
            orderRepository.deleteByShopId(id);
            
            // 3. Ledger entries delete karo
            ledgerRepository.deleteByShopId(id);
            
            // 4. Users mein se shop_id NULL karo
            userRepository.nullifyShopId(id.toString());
            
            // 5. Ab shop delete kar do
            shopRepository.deleteById(id);
            
            return ResponseEntity.ok("Shop deleted successfully!");
        } catch (Exception e) {
            return ResponseEntity.status(500).body("Error deleting shop: " + e.getMessage());
        }
    }

    // GET: Market Summary — Admin only
    @GetMapping("/market-summary")
    public ResponseEntity<?> getMarketSummary(@RequestAttribute("role") String role) {
        // ✅ SECURITY FIX: Only admin can see market summary
        if (!"admin".equalsIgnoreCase(role)) {
            return ResponseEntity.status(403).body("Access Denied: Only Admin can view market summary");
        }
        return ResponseEntity.ok(shopRepository.getMarketSummary());
    }
}

File: erp-backend-node/src/controllers/shopController.ts
import { Request, Response } from 'express';
import prisma from '../prisma/client';

// Get all shops
export const getShops = async (req: Request, res: Response) => {
  try {
    const shops = await prisma.shops.findMany({
      orderBy: { shop_name: 'asc' }
    });
    res.json(shops);
  } catch (error) {
    console.error('Error fetching shops:', error);
    res.status(500).json({ message: 'Server error while fetching shops.' });
  }
};

// Add a new shop
export const addShop = async (req: Request, res: Response) => {
  try {
    const { shop_name, shop_address } = req.body;
    const newShop = await prisma.shops.create({
      data: {
        shop_name,
        shop_address,
        total_debt: 0.00
      }
    });
    res.json(newShop);
  } catch (error) {
    console.error('Error adding shop:', error);
    res.status(500).json({ message: 'Server error while adding shop.' });
  }
};

// Delete a shop
export const deleteShop = async (req: Request, res: Response) => {
  try {
    const id = req.params.id as string;
    await prisma.shops.delete({
      where: { shop_id: parseInt(id) }
    });
    res.json({ message: 'Shop deleted successfully.' });
  } catch (error) {
    console.error('Error deleting shop:', error);
    res.status(500).json({ message: 'Server error while deleting shop.' });
  }
};

// Get market summary (total shops, total market debt)
export const getMarketSummary = async (req: Request, res: Response) => {
  try {
    const totalShops = await prisma.shops.count();
    const aggregate = await prisma.shops.aggregate({
      _sum: {
        total_debt: true
      }
    });
    
    res.json({
      totalShops,
      totalMarketDebt: aggregate._sum.total_debt || 0
    });
  } catch (error) {
    console.error('Error fetching market summary:', error);
    res.status(500).json({ message: 'Server error while fetching market summary.' });
  }
};
